### PR TITLE
add zones to cache metrics

### PR DESCRIPTION
**Pull request by elijahandrews**
This PR adds a `zones` label to the cache plugin's metrics. Currently, cache metrics are not tagged by zone; if you have multiple server blocks with the cache plugin enabled, the metrics for the different servers compete for last write (for gauges) or combine (for counters).

To reproduce this issue, consider the following Corefile:

```
(defaults) {
    prometheus 0.0.0.0:9253
    forward . /etc/resolv.conf
    cache
}

datadoghq.com cluster.local {
    import defaults
}

. {
    import defaults
}
```

After starting CoreDNS 1.8.6, make a few DNS queries to the three zones:

```
$ dig @127.0.0.1 +short datadoghq.com google.com doesnotexist.cluster.local
99.84.42.100
99.84.42.61
99.84.42.44
99.84.42.47
142.251.41.14

```

We can then curl the Prometheus endpoint for metrics emitted by the cache plugin:

```
$ curl -s localhost:9253/metrics | grep "^coredns_cache"
coredns_cache_entries{server="dns://:53",type="denial"} 1
coredns_cache_entries{server="dns://:53",type="success"} 1
coredns_cache_misses_total{server="dns://:53"} 3
coredns_cache_requests_total{server="dns://:53"} 3
```

For the counter metrics (`coredns_cache_misses_total` and `coredns_cache_requests_total`), the values are correct, but we cannot identify the values for the individual caches.

For the gauge metric (`coredns_cache_entries`), the values are incorrect, since the last write to the gauge wins. The sum of the denial and success cache should be 3, but reads 2. 

Following the same steps using the changes in this PR, we receive accurate measurements:

```
coredns_cache_entries{server="dns://:53",type="denial",zones="."} 0
coredns_cache_entries{server="dns://:53",type="denial",zones="datadoghq.com.,cluster.local."} 1
coredns_cache_entries{server="dns://:53",type="success",zones="."} 1
coredns_cache_entries{server="dns://:53",type="success",zones="datadoghq.com.,cluster.local."} 1
coredns_cache_misses_total{server="dns://:53",zones="."} 1
coredns_cache_misses_total{server="dns://:53",zones="datadoghq.com.,cluster.local."} 2
coredns_cache_requests_total{server="dns://:53",zones="."} 1
coredns_cache_requests_total{server="dns://:53",zones="datadoghq.com.,cluster.local."} 2
```

### CODE REVIEW

File: plugin/cache/cache.go [hunk 1]
```diff
@@ -162,8 +163,9 @@ func (w *ResponseWriter) WriteMsg(res *dns.Msg) error {
 	if hasKey && duration > 0 {
 		if w.state.Match(res) {
 			w.set(res, key, mt, duration)
-			cacheSize.WithLabelValues(w.server, Success).Set(float64(w.pcache.Len()))
-			cacheSize.WithLabelValues(w.server, Denial).Set(float64(w.ncache.Len()))
+			// TODO better way to do this without joins?
```
Comment (atollena) [inline plugin/cache/cache.go:166]: IIUC "Zones" is a hardcoded list (set here: https://github.com/coredns/coredns/blob/master/plugin/cache/setup.go#L187)

So I don't see any inherently better thing to do than using a separator (e.g. one gauge per zone doesn't sound right).

If you're concerned about performance of joining, perhaps you could compute this label only once on cache instantiation? (this requires doing that in setup.go which is a bit awkward I admit).

**Review by atollena — COMMENTED** [review]
Would it make sense to add this label for all cache metrics in https://github.com/coredns/coredns/blob/master/plugin/cache/metrics.go#L12-L66 ? (I think it would be useful)

File: plugin/cache/cache.go (hunk 1 above)
Comment (elijahandrews) [reply plugin/cache/cache.go]: addressed via https://github.com/DataDog/coredns/pull/69/commits/71f091455e71c105206c2144e41315b47a688ca6

**Comment by elijahandrews** [comment]
I'm trying to upstream this change [here](https://github.com/coredns/coredns/issues/5123)

File: plugin/cache/metrics.go [hunk 2]
```diff
@@ -14,47 +14,54 @@ var (
...
 	// cacheHits is counter of cache hits by cache type.
 	cacheHits = promauto.NewCounterVec(prometheus.CounterOpts{
 		Namespace: plugin.Namespace,
 		Subsystem: "cache",
 		Name:      "hits_total",
 		Help:      "The count of cache hits.",
-	}, []string{"server", "type"})
-	// cacheMisses is the counter of cache misses.
+	}, []string{"server", "type", "zones"})
+	// cacheMisses is the counter of cache misses. - Deprecated
```
Comment (atollena) [inline plugin/cache/metrics.go:32]: Did you take the initiative to deprecate this, or did you get that from somewhere else?

**Review by atollena — APPROVED** [review]

File: plugin/cache/metrics.go (hunk 2 above)
Comment (atollena) [reply plugin/cache/metrics.go:32]: OK, I see that it is already deprecated in the upstream repository.

Comment (elijahandrews) [reply plugin/cache/metrics.go:32]: oh yeah sorry, I just checked out these files from my branch that I made against upstream – this change came with it.

**Comment by atollena** [comment]
This has been merged upstream in https://github.com/coredns/coredns/pull/5124

**Pull request closed without merging**